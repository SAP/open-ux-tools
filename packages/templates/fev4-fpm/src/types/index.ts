export * from './page';

export interface Ui5Route {
    name: string;
    pattern: string;
    target: string | string[];
}
