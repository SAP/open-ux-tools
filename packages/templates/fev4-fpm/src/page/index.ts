import { join, dirname } from 'path';
import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';
import { render } from 'ejs';

import { CustomPage, CustomPageConfig } from '../types';

/**
 * Enhances the provided custom page configuration with default data
 *
 * @param {CustomPage} data - a custom page configuration object
 */
 export function enhanceData(data: CustomPage, manifestPath: string, fs: Editor): CustomPageConfig {
    // enforce naming conventions
    const firstChar = data.name[0];
    const nameForId = firstChar.toLocaleLowerCase() + data.name.substring(1);
    data.name = firstChar.toUpperCase() + data.name.substring(1)

    const manifest: any = fs.readJSON(manifestPath);
    const config: CustomPageConfig = {
        ...data,
        id: `${manifest['sap.app'].id}.ext.${nameForId}`,
        path: join(dirname(manifestPath), 'ext', nameForId)
    };
    if (config.view === undefined) {
        config.view = {
            title: config.name
        };
    }
    return config;
}

/**
 * Validates the provided base path.
 *
 * @param {string} basePath - the base path
 * @param {Editor} fs - the memfs editor instance
 */
export function validateBasePath(basePath: string, fs?: Editor) {
    if (!fs) {
        fs = create(createStorage());
    }

    const manifestPath = join(basePath, 'webapp', 'manifest.json');
    if (!fs.exists(manifestPath)) {
        throw new Error(`Invalid project folder. Cannot find required file ${manifestPath}`);
    } else {
        const manifest = fs.readJSON(manifestPath);
        // TODO: check if this is an FE app
    }

}
/**
 * Writes the template to the memfs editor instance.
 *
 * @param {string} basePath - the base path
 * @param {CustomPage} data - the custom page configuration
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {Promise<Editor>} the updated memfs editor instance
 */
export async function generateCustomPage(basePath: string, data: CustomPage, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    validateBasePath(basePath, fs);
    
    
    const manifestPath = join(basePath, 'webapp/manifest.json');
    const config = enhanceData(data, manifestPath, fs);
    
    // merge content into existing files
    const root = join(__dirname, '../../templates/page');

    // enhance manifest.json
    fs.extendJSON(manifestPath, JSON.parse(render(fs.read(join(root, `manifest.json`)), config)), ( key, value ) => {
        if (key === 'routes') {
            const routes = value as object[];
            routes.push({
                pattern: `${config.navigation ? config.navigation.sourceEntity + '({key})/' + config.navigation.navEntity + '({key2})' : config.entity + '({key})'}:?query:`,
                name: `${config.entity}${config.name}`,
                target: `${config.entity}${config.name}`
            });
        }
        return value;
    });

    // add extension content
    if (config.view?.path) {
        // TODO: copying is not that simple, controller path needs to be read from view.xml and if the controller is copied, it's id in the sources needs to be changed
        fs.copy(config.view.path, join(config.path, `${config.name}.view.xml`));
        fs.copy(config.view.path.replace('view.xml', 'controller.js'), join(config.path, `${config.name}.controller.js`));
    } else {
        fs.copyTpl(
            join(root,'ext/NAME/View.xml'), 
            join(config.path, `${config.name}.view.xml`), config);
        fs.copyTpl(
            join(root,'ext/NAME/Controller.js'), 
            join(config.path, `${config.name}.controller.js`), config);
    }

    return fs;
}