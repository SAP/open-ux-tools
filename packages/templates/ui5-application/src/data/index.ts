import { Ui5App } from './types';
import { appDefaults, packageDefaults, mergeUi5 } from './defaults';

export * from './types';

export function mergeWithDefaults(data: Ui5App): Ui5App {
    return {
        app: Object.assign(appDefaults(data.app.id), data.app),
        ui5: mergeUi5(data.ui5),
        package: Object.assign(packageDefaults(data.app.version), data.package)
    };
}
