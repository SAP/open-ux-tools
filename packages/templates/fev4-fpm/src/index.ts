import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';
import { render } from 'ejs';

import { CustomPage, CustomPageConfig, enhanceData } from './data';

/**
 * Validates the provided base path.
 *
 * @param {string} basePath - the base path
 * @param {Editor} fs - the memfs editor instance
 */
function validateBasePath(basePath: string, fs: Editor) {
    [join(basePath, 'webapp', 'manifest.json')].forEach(
        (path) => {
            if (!fs.exists(path)) {
                throw new Error(`Invalid project folder. Cannot find required file ${path}`);
            }
        }
    );
}
/**
 * Writes the template to the memfs editor instance.
 *
 * @param {string} basePath - the base path
 * @param {CustomPage} data - the custom page configuration
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {Promise<Editor>} the updated memfs editor instance
 */
async function generateCustomPage(basePath: string, data: CustomPage, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    validateBasePath(basePath, fs);
    
    
    const manifestPath = join(basePath, 'webapp/manifest.json');
    const config = enhanceData(data, manifestPath, fs);
    
    // merge content into existing files
    const root = join(__dirname, '..', 'templates/customPage');

    // enhance manifest.json
    fs.extendJSON(manifestPath, JSON.parse(render(fs.read(join(root, `manifest.json`)), config)));

    // add extension content
    if (data.view.path === undefined) {
        data.view.path = join(basePath, `webapp/ext/view/${config.view.name}.view.xml`);
        fs.copyTpl(
            join(root,'ext/view/CustomPage.view.xml'), 
            data.view.path, config);
        fs.copyTpl(
            join(root,'ext/controller/CustomPage.controller.js'), 
            join(basePath, `webapp/ext/controller/${config.view.name}.controller.js`), config);
    }

    return fs;
}

export { CustomPage, generateCustomPage};
