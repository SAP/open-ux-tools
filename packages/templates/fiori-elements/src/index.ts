import { join } from 'path';
import { Editor } from 'mem-fs-editor';
import { render } from 'ejs';

import { generate as generateUi5Project } from '@sap/ux-ui5-application-template';
import { generate as addOdataService } from '@sap/ux-odata-service-template';
import { FEApp, getBaseComponent } from './data';
import { UI5Config } from '../../../lib/ui5-config/dist';
import { getUI5Libs } from './data/ui5Libs';
import { getMiddlewareConfig } from './data/middlewareConfig';

/**
 * @param basePath
 * @param data
 * @param fs
 */
async function generate<T>(basePath: string, data: FEApp<T>, fs?: Editor): Promise<Editor> {
    // generate base UI5 project
    data.app.baseComponent = getBaseComponent(data.template);
    fs = await generateUi5Project(basePath, data, fs);

    // add new and overwrite files from templates e.g. annotations.xml
    const tmpPath = join(__dirname, '..', 'templates', 'add');
    fs.copyTpl(join(tmpPath, '**/*.*'), basePath, data);

    // merge content into existing files
    const extRoot = join(__dirname, '..', 'templates', 'extend');

    // package.json
    const packagePath = join(basePath, 'package.json');
    fs.extendJSON(packagePath, fs.readJSON(join(extRoot, 'package.json')));
    const packageJson = JSON.parse(fs.read(packagePath));
    packageJson.ui5.dependencies.push('@sap/ux-ui5-tooling');
    fs.writeJSON(packagePath, packageJson);

    // manifest.json
    const manifestPath = join(basePath, 'webapp', 'manifest.json');
    fs.extendJSON(manifestPath, fs.readJSON(join(extRoot, 'manifest.json')));
    fs.extendJSON(
        manifestPath,
        JSON.parse(render(fs.read(join(extRoot, data.template.version, `manifest.${data.template.type}.json`)), data))
    );

    // ui5.yaml
    const ui5ConfigPath = join(basePath, 'ui5.yaml');
    const ui5Config = await UI5Config.newInstance(fs.read(ui5ConfigPath));
    ui5Config
        .addLibraries(getUI5Libs(data.template.type, data.template.version))
        .addCustomMiddleware(getMiddlewareConfig());
    fs.write(ui5ConfigPath, ui5Config.toString());

    // add service to the project
    await addOdataService(basePath, data.service, fs);
    const manifest = JSON.parse(fs.read(manifestPath));
    manifest['sap.app'].dataSources[data.service.name!].settings.annotations.push('annotation');
    fs.writeJSON(manifestPath, manifest);
    return fs;
}

export { generate };
