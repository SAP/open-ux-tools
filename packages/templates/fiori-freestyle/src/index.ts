import { join } from 'path';
import { Editor } from 'mem-fs-editor';
import { render } from 'ejs';

import { generate as generateUi5Project } from '@sap/ux-ui5-application-template';
import { generate as addOdataService } from '@sap/ux-odata-service-template';
import { FreestyleApp, WorklistSettings, ListDetailSettings, TemplateType } from './data';
import { UI5Config, getAppReloadMiddlewareConfig } from '@sap/ux-ui5-config';
import { getUI5Libs } from './data/ui5Libs';

/**
 * Generate a UI5 application based on the specified Fiori Freestyle floorplan template.
 *
 * @param basePath - the absolute target path where the applciation will be generated
 * @param data - ???
 * @param fs - an optional reference to a mem-fs editor
 * @returns Reference to a mem-fs-editor
 */
async function generate<T>(basePath: string, data: FreestyleApp<T>, fs?: Editor): Promise<Editor> {
    // generate base UI5 project
    data.app.baseComponent = 'sap/ui/core/UIComponent';
    fs = await generateUi5Project(basePath, data, fs);

    // add new and overwrite files from templates e.g.
    const tmpPath = join(__dirname, '..', 'templates');
    // Common files
    fs.copyTpl(join(tmpPath, 'common', 'add', '**/*.*'), basePath, data);

    // By template type
    fs.copyTpl(join(tmpPath, data.template.type, 'add', '**/*.*'), basePath, data);

    // merge content into existing files
    const extRoot = join(__dirname, '..', 'templates', data.template.type, 'extend', 'webapp');

    // manifest.json
    const manifestPath = join(basePath, 'webapp', 'manifest.json');
    fs.extendJSON(manifestPath, JSON.parse(render(fs.read(join(extRoot, 'manifest.json')), data)));

    // i18n.properties
    fs.append(
        join(basePath, 'webapp', 'i18n', 'i18n.properties'),
        render(fs.read(join(extRoot, 'i18n', 'i18n.properties')), data)
    );

    // package.json
    const packagePath = join(basePath, 'package.json');
    fs.extendJSON(packagePath, JSON.parse(render(fs.read(join(tmpPath, 'common', 'extend', 'package.json')), data)));
    const packageJson = JSON.parse(fs.read(packagePath));
    packageJson.ui5.dependencies.push('@sap/ux-ui5-tooling');
    fs.writeJSON(packagePath, packageJson);

    // ui5.yaml
    const ui5ConfigPath = join(basePath, 'ui5.yaml');
    const ui5Config = await UI5Config.newInstance(fs.read(ui5ConfigPath));
    ui5Config.addCustomMiddleware(getAppReloadMiddlewareConfig());
    fs.write(ui5ConfigPath, ui5Config.toString());

    // add service to the project if provided
    if (data.service) {
        await addOdataService(basePath, data.service, fs);
    }

    // ui5-local.yaml
    const ui5LocalConfigPath = join(basePath, 'ui5-local.yaml');
    const ui5LocalConfig = await UI5Config.newInstance(fs.read(ui5LocalConfigPath));
    if (data?.ui5?.localVersion) {
        ui5Config.addUI5Framework(data.ui5.localVersion, getUI5Libs(data?.ui5?.ui5Libs));
    }
    ui5LocalConfig.addCustomMiddleware(getAppReloadMiddlewareConfig());

    fs.write(ui5LocalConfigPath, ui5Config.toString());

    return fs;
}

export { generate, FreestyleApp, WorklistSettings, ListDetailSettings, TemplateType };
