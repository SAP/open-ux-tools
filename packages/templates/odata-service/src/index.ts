import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';
import { render } from 'ejs';
import { UI5Config } from '@sap/ux-ui5-config';

import { OdataService, OdataVersion, enhanceData } from './data';
import { getMiddlewareConfig, getLocalMiddlewareConfig } from './data/middleware';

/**
 * Validates the provided base path.
 *
 * @param {string} basePath - the base path
 * @param {Editor} fs - the memfs editor instance
 */
function validateBasePath(basePath: string, fs: Editor) {
    [join(basePath, 'package.json'), join(basePath, 'webapp', 'manifest.json'), join(basePath, 'ui5.yaml')].forEach(
        (path) => {
            if (!fs.exists(path)) {
                throw new Error(`Invalid project folder. Cannot find required file ${path}`);
            }
        }
    );
}
/**
 * Writes the template to the memfs editor instance.
 *
 * @param {string} basePath - the base path
 * @param {OdataService} data - the OData service instance
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {Promise<Editor>} the updated memfs editor instance
 */
async function generate(basePath: string, data: OdataService, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    validateBasePath(basePath, fs);
    enhanceData(data);

    // add new and overwrite files from templates
    //const tmpPath = join(__dirname, 'templates', 'add');
    //fs.copyTpl(join(tmpPath, '**/*.*'), basePath, data);

    // merge content into existing files
    const extRoot = join(__dirname, '..', 'templates', 'extend');

    // package.json
    const packagePath = join(basePath, 'package.json');
    fs.extendJSON(packagePath, fs.readJSON(join(extRoot, 'package.json')));
    const packageJson = JSON.parse(fs.read(packagePath));
    packageJson.ui5.dependencies.push('@sap/ux-ui5-fe-mockserver-middleware');
    fs.writeJSON(packagePath, packageJson);

    // manifest.json
    const manifestPath = join(basePath, 'webapp', 'manifest.json');
    fs.extendJSON(manifestPath, JSON.parse(render(fs.read(join(extRoot, `manifest.json`)), data)));

    // ui5.yaml and ui5-local.yaml
    const ui5ConfigPath = join(basePath, 'ui5.yaml');
    const existingUI5Config = fs.read(ui5ConfigPath);

    const ui5Config = await UI5Config.newInstance(existingUI5Config);
    ui5Config.addCustomMiddleware(...getMiddlewareConfig(data));
    fs.write(ui5ConfigPath, ui5Config.toString());

    const ui5LocalConfig = await UI5Config.newInstance(existingUI5Config);
    ui5LocalConfig.addCustomMiddleware(getLocalMiddlewareConfig(data));
    fs.write(join(basePath, 'ui5-local.yaml'), ui5LocalConfig.toString());

    // create local copy of metadata and annotations
    if (data.metadata) {
        fs.write(join(basePath, 'webapp', 'localService', 'metadata.xml'), data.metadata);
    }
    if (data.annotations?.xml) {
        fs.write(
            join(basePath, 'webapp', 'localService', `${data.annotations.technicalName}.xml`),
            data.annotations.xml
        );
    }

    return fs;
}

export { generate, OdataVersion, OdataService };
