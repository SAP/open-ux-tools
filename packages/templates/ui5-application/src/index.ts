import { join } from 'path';
import { Ui5App, mergeWithDefaults } from './data';
import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';

/**
 * Writes the template to the memfs editor instance.
 *
 * @param {string} basePath - the base path
 * @param {Ui5App} data - the Ui5App instance
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {*}  {Promise<Editor>} the updated memfs editor instance
 */
async function generate(basePath: string, data: Ui5App, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    const fullData = mergeWithDefaults(data);
    const tmpPath = join(__dirname, '..', 'templates');

    fs.copyTpl(join(tmpPath, '**/*.*'), join(basePath), fullData);

    return fs;
}

export { Ui5App, generate };
