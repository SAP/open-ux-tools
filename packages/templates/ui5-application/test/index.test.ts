import { generate } from '../src';
import { join } from 'path';
import { tmpdir } from 'os';
import { rmdirSync } from 'fs';

describe('UI5 templates', () => {
    const debug = !!process.env['UX_DEBUG'];
    const outputDir = join(tmpdir(), '/templates/ui5-application');
    if (debug) console.log(outputDir);

    afterEach(() => {
        if (!debug) rmdirSync(outputDir, { recursive: true });
    });

    it('generates files correctly', async () => {
        const fs = await generate(outputDir, {
            app: {
                id: 'myApp',
                title: 'My application'
            },
            package: {
                name: 'Test'
            }
        });
        if (debug) fs.commit(() => 0);
        expect((fs as any).dump(outputDir)).toMatchSnapshot();
    });
});
