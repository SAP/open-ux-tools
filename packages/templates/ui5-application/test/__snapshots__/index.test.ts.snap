// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI5 templates generates files correctly 1`] = `
Object {
  "package.json": Object {
    "contents": "{
	\\"name\\": \\"test\\",
	\\"version\\": \\"0.9.0\\",
	\\"main\\": \\"dist/index.html\\",
	\\"dependencies\\": {},
	\\"devDependencies\\": {
    \\"@ui5/cli\\": \\"^2.12.0\\"
},
	\\"scripts\\": {
    \\"start\\": \\"ui5 serve --config=ui5.yaml --open index.html\\",
    \\"build\\": \\"ui5 build --config=ui5.yaml --clean-dest --dest dist\\"
},
	\\"ui5\\": {
		\\"dependencies\\": []
	},
	\\"files\\": [
		\\"dist\\"
	]
}",
    "state": "modified",
  },
  "ui5.yaml": Object {
    "contents": "specVersion: \\"2.5\\"
metadata:
  name: myApp
type: application
resources:
  configuration:
    propertiesFileSourceEncoding: UTF-8
framework:
  name: SAPUI5
  version: 1.84.0
  libraries:
    - name: sap.ui.core
    - name: sap.m
    - name: sap.ui.layout
    - name: themelib_sap_fiori_3
",
    "state": "modified",
  },
  "webapp/Component.js": Object {
    "contents": "sap.ui.define(
    [\\"sap/ui/core/UIComponent\\"],
    function (UIComponent) {
        \\"use strict\\";

        return UIComponent.extend(\\"myApp.Component\\", {
            metadata: {
                manifest: \\"json\\"
            },

            /**
             * The component is initialized by UI5 automatically during the startup of the app and calls the init method once.
             * @public
             * @override
             */
            init: function () {
                // call the base component's init function
                UIComponent.prototype.init.apply(this, arguments);

                // enable routing
                this.getRouter().initialize();
            }
        });
    }
);",
    "state": "modified",
  },
  "webapp/i18n/i18n.properties": Object {
    "contents": "#XTIT: Application name
appTitle=My application

#YDES: Application description
appDescription=Description of myApp",
    "state": "modified",
  },
  "webapp/i18n/i18n_en.properties": Object {
    "contents": "appTitle=My application
appDescription=Description of myApp",
    "state": "modified",
  },
  "webapp/index.html": Object {
    "contents": "<!DOCTYPE html>
<html>
<head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <title>My application</title>
    <style>
        html, body, body > div, #container, #container-uiarea {
            height: 100%;
        }
    </style>
    <script
        id=\\"sap-ui-bootstrap\\"
        src=\\"resources/sap-ui-core.js\\"
        data-sap-ui-theme=\\"sap_fiori_3\\"
        data-sap-ui-resourceroots='{
            \\"myApp\\": \\"./\\"
        }'
        data-sap-ui-oninit=\\"module:sap/ui/core/ComponentSupport\\"
        data-sap-ui-compatVersion=\\"edge\\"
        data-sap-ui-async=\\"true\\"
        data-sap-ui-frameOptions=\\"trusted\\"
    ></script>
</head>
<body class=\\"sapUiBody sapUiSizeCompact\\" id=\\"content\\">
    <div
        data-sap-ui-component
        data-name=\\"myApp\\"
        data-id=\\"container\\"
        data-settings='{\\"id\\" : \\"myApp\\"}'
        data-handle-validation=\\"true\\"
    ></div>
</body>
</html>",
    "state": "modified",
  },
  "webapp/manifest.json": Object {
    "contents": "{
    \\"_version\\": \\"1.12.0\\",
    \\"sap.app\\": {
        \\"id\\": \\"myApp\\",
      \\"type\\": \\"application\\",
      \\"i18n\\": \\"i18n/i18n.properties\\",
      \\"applicationVersion\\": {
        \\"version\\": \\"0.9.0\\"
      },
      \\"title\\": \\"{{appTitle}}\\",
      \\"description\\": \\"{{appDescription}}\\"
    },
    \\"sap.ui\\": {
      \\"technology\\": \\"UI5\\",
      \\"icons\\": {
        \\"icon\\": \\"\\",
        \\"favIcon\\": \\"\\",
        \\"phone\\": \\"\\",
        \\"phone@2\\": \\"\\",
        \\"tablet\\": \\"\\",
        \\"tablet@2\\": \\"\\"
      },
      \\"deviceTypes\\": {
        \\"desktop\\": true,
        \\"tablet\\": true,
        \\"phone\\": true
      }
    },
    \\"sap.ui5\\": {
      \\"flexEnabled\\": true,
      \\"dependencies\\": {
        \\"minUI5Version\\": \\"1.60\\",
        \\"libs\\": {
          \\"sap.ui.core\\": {}
        }
      },
      \\"contentDensities\\": {
        \\"compact\\": true,
        \\"cozy\\": true
      },
      \\"models\\": {
        \\"i18n\\": {
          \\"type\\": \\"sap.ui.model.resource.ResourceModel\\",
          \\"settings\\": {
            \\"bundleName\\": \\"myApp.i18n.i18n\\"
          }
        }
      },
      \\"resources\\": {
        \\"css\\": []
      },
      \\"routing\\": {
        \\"config\\": {},
        \\"routes\\": [],
        \\"targets\\": {}
      }
    }
  }",
    "state": "modified",
  },
}
`;
